pkg/volume/quobyte/quobyte.go

```diff
@@ -0,0 +1,243 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package quobyte
+
+import (
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// ProbeVolumePlugins is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&quobytePlugin{nil, exec.New()}}
+}
+
+type quobytePlugin struct {
+	host volume.VolumeHost
+	exe  exec.Interface
+}
+
+var _ volume.VolumePlugin = &quobytePlugin{}
+var _ volume.PersistentVolumePlugin = &quobytePlugin{}
+
+const (
+	quobytePluginName = "kubernetes.io/quobyte"
+)
+
+func (plugin *quobytePlugin) Init(host volume.VolumeHost) error {
+	plugin.host = host
+	return nil
+}
+
+func (plugin *quobytePlugin) Name() string {
+	return quobytePluginName
+}
+
+func (plugin *quobytePlugin) CanSupport(spec *volume.Spec) bool {
+	if (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Quobyte == nil) ||
+		(spec.Volume != nil && spec.Volume.Quobyte == nil) {
+		return false
+	}
+	if _, err := plugin.execCommand("ls", []string{"/sbin/mount.quobyte"}); err == nil {
+		return true
+	}
+
+	return false
+}
+
+func (plugin *quobytePlugin) GetAccessModes() []api.PersistentVolumeAccessMode {
+	return []api.PersistentVolumeAccessMode{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+		api.ReadWriteMany,
+	}
+}
+
+func (plugin *quobytePlugin) NewMounter(spec *volume.Spec, pod *api.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
+	source, _ := plugin.getQuobyteVolumeSource(spec)
+	registry := source.Registry
+	glog.V(1).Infof("quobyte: registry %s", registry)
+	return plugin.newMounterInternal(spec, registry, pod, plugin.host.GetMounter(), exec.New())
+}
+
+func (plugin *quobytePlugin) getQuobyteVolumeSource(spec *volume.Spec) (*api.QuobyteVolumeSource, bool) {
+	// Quobyte volumes used directly in a pod have a ReadOnly flag set by the pod author.
+	// Quobyte volumes used as a PersistentVolume gets the ReadOnly flag indirectly through the persistent-claim volume used to mount the PV
+	if spec.Volume != nil && spec.Volume.Quobyte != nil {
+		return spec.Volume.Quobyte, spec.Volume.Quobyte.ReadOnly
+	}
+
+	return spec.PersistentVolume.Spec.Quobyte, spec.ReadOnly
+}
+
+func (plugin *quobytePlugin) newMounterInternal(spec *volume.Spec, registry string, pod *api.Pod, mounter mount.Interface, exe exec.Interface) (volume.Mounter, error) {
+	source, readOnly := plugin.getQuobyteVolumeSource(spec)
+	return &quobyteMounter{
+		quobyte: &quobyte{
+			volName: spec.Name(),
+			mounter: mounter,
+			pod:     pod,
+			plugin:  plugin,
+		},
+		registry: registry,
+		volume:   source.Volume,
+		path:     source.Path,
```

Comment:
@childsb Quobyte engineering will maintain it. Generally this is not complicated stuff as it is just a mapping that decouples Pod descriptions from the actual setup of the underlying infrastructure. Most changes should be guarded by the tests (that's what they are for in the end), if it still breaks, we'll fix it.

We looked at Flexvolume but it declares itself "alpha". As people want to use this in production, Flexvolume didn't seem the way to go.